Add annotation support and HtmxResponse methods for all existing htmx response headers

xhaggi opened the pull request:
This adds annotation support and `HtmxResponse` methods for all currently existing [htmx Response Headers](https://htmx.org/reference/#response_headers). It also introduces a builder in `HtmxResponse`.

Please check the respective commits.

wimdeblauwe commented on htmx-spring-boot/src/main/java/io/github/wimdeblauwe/htmx/spring/boot/mvc/HxPushUrl.java:
```diff
@@ -0,0 +1,24 @@
+    /**
+     * The URL to be pushed into the location bar. This may be relative or absolute,
+     * as per <a href="https://developer.mozilla.org/en-US/docs/Web/API/History/pushState">history.pushState()</a>.
+     * Or {@code false}, which prevents the browser’s history from being updated.
```
I wonder if we should have a constant somewhere to represent the `"false"` string, as it might not be obvious you need to return false as a String  literal.

xhaggi replied:
Sure, how about `HtmxValue.FALSE`?

wimdeblauwe commented on htmx-spring-boot/src/main/java/io/github/wimdeblauwe/htmx/spring/boot/mvc/HxReswap.java:
```diff
@@ -0,0 +1,25 @@
+     *
+     * @see <a href="https://htmx.org/attributes/hx-swap/">hx-swap</a>
+     */
+    String value();
```
Should we return an enum value here to make it easier to select the correct values?

xhaggi replied:
The value of `HX-Reswap` supports the same options as the [hx-swap](https://htmx.org/attributes/hx-swap/) attribute. So, in my opinion, the current implementation that only supports the values of the `HxSwapType` enum in `HtmxResponse` is wrong. I would fix it by adding a class that also supports all modifier options. Let me just do that for a second.

xhaggi commented:
I'm done, probably a little more 😄.  I added an additional commit `Add support for specifying modifiers in HX-Reswap` which changes the implementation to support all modifiers via the `HtmxReswap` class. I updated the commit `Add annotation support for HX-Reswap` to support all modifiers in the annotation as well. 

```java
@HxReswap(value = HxSwapType.INNER_HTML, swap = 300)
```
vs.
```java
return HtmxResource.builder()
  .reswap(HtmxReswap.innerHtml().swap(Duration.ofMillis(300)))
  .build();
```

I also rewrote the HX-Trigger implementation in `HtmxResponse` to support complex event details via `Map`. To simplify things a bit, I introduced a new class `HtmxTrigger` and dedicated builder methods for each case.

```java
return HtmxResource.builder()
  .trigger("event1")
  .trigger("event2", Map.of("var", "value"))
  .triggerAfterSettle("event3")
  .triggerAfterSettle("event4", Map.of("var", "value"))
  .triggerAfterSwap("event5")
  .triggerAfterSwap("event6", Map.of("var", "value"))
  .build();
```

It would really make sense to start a fresh review to not miss anything.

xhaggi commented:
I have added `HxSwapType.INNER_HTML` as a default value in the `HxReswap` annotation, so there is no need to redefine this if you just want to customize a modifier e.g. `swap`.

wimdeblauwe commented on htmx-spring-boot/src/main/java/io/github/wimdeblauwe/htmx/spring/boot/mvc/HxReswap.java:
```diff
@@ -0,0 +1,87 @@
+    /**
+     * Set the time in milliseconds that should elapse after receiving a response to swap the content.
+     */
+    long swap() default -1;
```
I am wondering if we should use `Duration` here?

xhaggi replied:
You can't use Duration here. See [9.6.1. Annotation Interface Elements](https://docs.oracle.com/javase/specs/jls/se17/html/jls-9.html#jls-9.6.1)

>The return type of a method declared in the body of annotation interface must be one of the following, or a compile-time error occurs:
>
>- A primitive type
>- String
>- Class or an invocation of Class ([§4.5](https://docs.oracle.com/javase/specs/jls/se17/html/jls-4.html#jls-4.5))
>- An enum class type
>- An annotation interface type
>- An array type whose component type is one of the preceding types ([§10.1](https://docs.oracle.com/javase/specs/jls/se17/html/jls-10.html#jls-10.1)).

wimdeblauwe replied:
Ah, ok, that is a pity.

wimdeblauwe commented on htmx-spring-boot/src/main/java/io/github/wimdeblauwe/htmx/spring/boot/mvc/HtmxTrigger.java:
```diff
@@ -0,0 +1,46 @@
+ *
+ * @see <a href="https://htmx.org/headers/hx-trigger/">HX-Trigger Response Headers</a>
+ */
+public class HtmxTrigger {
```
This could be a record I think since it is immuable?

xhaggi replied:
I am not sure if we should really use a record. These are not very flexible with regard to any future adjustments.

xhaggi commented:
I have added 2 more commits to update the README.md. I hope it's ok that I added the [javadoc.io](https://javadoc.io/) badge and link part of the README to javadoc.io.

xhaggi commented:
@wimdeblauwe just wanted to ask if anything else is missing? Or is there something else to discuss besides the open comments?

wimdeblauwe commented:
Thank you for your work so far, but I wonder if we can split up the PR into multiple PR's? Because it is really big as it is now and I am loosing a bit the full picture like this. If possible, I would like to have separate PR's for:

1. Adding the javadoc.io badge and linking to specific pages from the readme file.
2. Reworking `HtmxResponse` to use the builder pattern (Not sure if we should do this, so having this in a separate PR makes it easier to ask for other people's opinion on this).
3.  `HtmxRequest` renames of methods to remove the get/set prefix. If we do this, we should keep the old methods and deprecate them first to avoid people getting compilation errors on upgrading (I would like to avoid bumping the major version for this, again not sure if this brings much value to the project)
4. Add Javadoc to existing code. This can be a straighforward addition and a PR I can merge without any problem of having breaking changes.
5. Additional annotations. Ideally, we could have a separate PR per annotation, but I understand that would be a lot of work and might be a bit easier to group those together in 1 PR.

If you see other ways to break up this big PR, I am also open for discussion on this.

xhaggi commented:
I don't want to spend too much time splitting it up, since there is no real advantage and many of the commits rely on each other. I definitely understand your concerns about introducing the builder pattern for `HtmxResponse`. My intention here was to make the implementation more like existing classes such as [ResponseEntity](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/http/ResponseEntity.html), which also use a builder pattern (without `with` prefix). Therefore, I would suggest to keep the old implementation, mark it as deprecated and refer to the new builder pattern. The changes to `HtmxRequest` should be fairly safe, since `HtmxRequest` is not or should not be built by yourself. We could keep the changes or we could mark the methods as deprecated, as you like?

In my opinion, I would then split up the following commits into separate PRs.

1. Wrong response header HX-Push instead of HX-Push-Url used
2. Add missing Javadoc + Add javadoc.io badge to README

Let me know what we want to do.

wimdeblauwe commented:
@checketts @odrotbohm I would like to hear what you think on this.

checketts commented:
I second the idea of splitting these changes into several smaller ones. If they truly need to build on each other, perhaps creating the PRs one at a time will be more useful.

If the original author isn't able to split them up, a separate issue for each effort, discussing the change would be another option, but won't solve the ability to merge in changes as things get resolved.

I don't mean to make extra work for @xhaggi just trying to approach this in a way that we can make progress bit by bit.

xhaggi commented:
Ok, let's start with the first round of splitting things up. There are now 4 new PR's to review. After these are merged, I will create another one that introduces the builder pattern into `HtmxResponse` and keeps backward compatibility.

checketts commented:
Thanks @xhaggi that made it much easier to review. Thanks for your contributions!

wimdeblauwe commented:
@xhaggi I suppose I can close this PR now as we have the others now?

xhaggi commented:
If that's OK with you, I'd like to leave this open until all the missing commits are in separate PRs.

* [Add support for specifying modifiers in HX-Reswap](https://github.com/wimdeblauwe/htmx-spring-boot/pull/67/commits/5a72afdfd60c6e9b8ab233c963bbccd313d12c31)
* [Add support for HX-Location response header](https://github.com/wimdeblauwe/htmx-spring-boot/pull/67/commits/8ab4c8997dd852c5faec86e4d74fb59f5facf725)
* [Add support for HX-Replace-Url response header](https://github.com/wimdeblauwe/htmx-spring-boot/pull/67/commits/883cd95f9e62aee0c77a92c9510df9318d25dc74)
* [Add support for HX-Reselect response header](https://github.com/wimdeblauwe/htmx-spring-boot/pull/67/commits/68d85a0e1c19ffd2c0a8ab1d830c647c3fdaf4f5)

xhaggi commented:
BTW you can [convert the pull request to a draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request#converting-a-pull-request-to-a-draft).

xhaggi commented:
@wimdeblauwe, we have nearly everything merged, expect the annotation support, which is something for 4.0. IMO you can now close this PR. And after the 2 open pull requests have been merged, we could finally release version 3.1.

wimdeblauwe commented:
> @wimdeblauwe, we have nearly everything merged, expect the annotation support, which is something for 4.0. IMO you can now close this PR. And after the 2 open pull requests have been merged, we could finally release version 3.1.

ok. I will close this PR. You have done amazing work on this version, thank you very, very much!
